### PR TITLE
Feat: add support for LPG

### LINKED ISSUE #38: add Local Peering Gateway support
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

VCN module currently supports creating each available gateways, except for [Local Peering Gateways](https://registry.terraform.io/providers/hashicorp/oci/latest/docs/resources/core_local_peering_gateway) (referred as LPG now on).

The proposition is to give the user an **_option_** to create one or many LPG within the same VCN block definition, with the following options:
- declare its [`peer_id`](https://registry.terraform.io/providers/hashicorp/oci/latest/docs/resources/core_local_peering_gateway#peer_id) or not,
- attach a Route Table to the LPG or not, by setting [`route_table_id`](https://registry.terraform.io/providers/hashicorp/oci/latest/docs/resources/core_local_peering_gateway#route_table_id)

_Adding one or many LPG to a VCN should be kept optional._

These options will enable higher level constructs, like hub-and-spoke network topology. But it is important to keep VCN module agnostic to such specific configuration:
- based on the arguments passed by the user to the VCN module, the resulting VCN should be able to act either as a HUB or a SPOKE,
- when composing with many VCN modules, the user should be able to create a complete hub-and-spoke topology.

### New or Affected Resource(s)

- `oci_core_vcn` will need new Input Variables
- new resource `oci_core_local_peering_gateway`, implemented with a `for_each` logic to handle many of them,
- possibly new `oci_route_table` to handle HUB scenario (preferably via a sub-module for route tables)

### Potential Terraform Configuration

Below a mock-up of how the user would declare new LPGs 

```hcl
module "vcn" {
  source = "../"

  # general oci parameters
  compartment_id = var.compartment_id
  label_prefix   = var.label_prefix
  tags           = var.tags

  # vcn parameters
  create_drg               = var.create_drg               # boolean: true or false
  internet_gateway_enabled = var.internet_gateway_enabled # boolean: true or false
  lockdown_default_seclist = var.lockdown_default_seclist # boolean: true or false
  nat_gateway_enabled      = var.nat_gateway_enabled      # boolean: true or false
  service_gateway_enabled  = var.service_gateway_enabled  # boolean: true or false
  vcn_cidr                 = var.vcn_cidr                 # VCN CIDR
  vcn_dns_label            = var.vcn_dns_label
  vcn_name                 = var.vcn_name

  # gateways parameters
  drg_display_name = var.drg_display_name

  local_peering_gateways = {
    to_spoke1 = {
      peer_id = <string>
      route_table_id = <string>
    },
    to_spoke2 = {
      peer_id = <string>
      route_table_id = <string>
    },
  }
}
```

### PR DESCRIPTION
**Pull request by kral2**
Fix #38 

- each vcn can have one or many lpg
- peer_id argument is supported and optional
- route_table_id is supported and optional

new example using this feature to build hub-and-spoke.

### CODE REVIEW

File: examples/hub-spoke/variables.tf [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+#     {
+#       destination       = "203.0.113.0/24" # rfc5737 (TEST-NET-3)
+#       destination_type  = "CIDR_BLOCK"
+#       network_entity_id = "nat_gateway"
+#       description       = "Terraformed - User added Routing Rule: To NAT Gateway created by this module. nat_gateway_id is automatically retrieved with keyword nat_gateway"
+#     },
+#     {
+#       destination       = "192.168.1.0/24"
+#       destination_type  = "CIDR_BLOCK"
+#       network_entity_id = oci_core_local_peering_gateway.lpg.id
```
Comment (hyder) [inline examples/hub-spoke/variables.tf:153]: This additional lpg example needs to be the in root terraform.tfvars.example too.

File: examples/hub-spoke/main.tf [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  lockdown_default_seclist = true          # boolean: true or false
+  nat_gateway_enabled      = false         # boolean: true or false
+  service_gateway_enabled  = false         # boolean: true or false
+  vcn_cidr                 = "10.0.1.0/24" # VCN CIDR
+  vcn_dns_label            = "fraspoke1"
+  vcn_name                 = "spoke1"
+
+  # gateways parameters
+
+  local_peering_gateways = {
```
Comment (hyder) [inline examples/hub-spoke/main.tf:77]: add commented example in root terraform.tfvars.example

**Review by hyder — CHANGES_REQUESTED** [review]
In a previous release, you also added a drg attachment : https://github.com/oracle-terraform-modules/terraform-oci-vcn/blob/main/vcn_gateways.tf#L207. See also issue https://github.com/oracle-terraform-modules/terraform-oci-vcn/issues/44

The problem is the following:

1. A VCN can have only 1 DRG at a time but a DRG can have multiple attachment.
2. The DRG attachment that is currently created in the VCN attaches the VCN to this DRG. However, we may need to specify additional parameters such as network details: See here: https://registry.terraform.io/providers/hashicorp/oci/latest/docs/resources/core_drg_attachment
3. This is particularly relevant if have a firewall in the hub and we therefore need to specify a private IP address in a custom route table and attach this route table to the drg_attachment.

The drg_attachment should therefore be able to accept custom route table ids. Currently, it does not.

**Comment by kral2** [comment]
> In a previous release, you also added a drg attachment : https://github.com/oracle-terraform-modules/terraform-oci-vcn/blob/main/vcn_gateways.tf#L207. See also issue #44
> 
> The problem is the following:
> 
>     1. A VCN can have only 1 DRG at a time but a DRG can have multiple attachment.
> 
>     2. The DRG attachment that is currently created in the VCN attaches the VCN to this DRG. However, we may need to specify additional parameters such as network details: See here: https://registry.terraform.io/providers/hashicorp/oci/latest/docs/resources/core_drg_attachment
> 
>     3. This is particularly relevant if have a firewall in the hub and we therefore need to specify a private IP address in a custom route table and attach this route table to the drg_attachment.
> 
> 
> The drg_attachment should therefore be able to accept custom route table ids. Currently, it does not.

This PR is related to issue #38 only: creating and controlling LPGs. The DRG attachment problem is related to the new capabilities of the DRG and should be handled in a different issue.

File: vcn_gateways.tf [hunk 3]
```diff
@@ -210,3 +210,21 @@ resource "oci_core_drg_attachment" "drg" {
 
   count = var.create_drg == true ? 1 : 0
 }
+
```
Comment (slmjy) [inline vcn_gateways.tf:213]: I suppose LPG support would be incomplte without internet and nad gateway rules, same as with DRG. So I suggest a code like this for **oci_core_route_table.nat** and **oci_core_route_table.ig**
```terraform
  dynamic "route_rules" {
    # * filter var.nat_gateway_route_rules for routes with "lpg[lpg_KEY]" as destination
    # * and steer traffic to the module created LPG with given key
    for_each = var.nat_gateway_route_rules != null ? { for k, v in var.nat_gateway_route_rules : k => v
    if v.network_entity_id == "lpg" } : {}

    content {
      destination       = route_rules.value.destination
      destination_type  = route_rules.value.destination_type
      network_entity_id = oci_core_local_peering_gateway.lpg[route_rules.value.lpg_key].id
      description       = route_rules.value.description
    }
  }
```

Comment (kral2) [reply vcn_gateways.tf:213]: Hello Andrey,

Thanks for bringing this point to my attention.

The idea here is to focus on a simple task (create LPGs) and to be as close possible to the provider capability. What's important, is to be able to attach (any) route table, whether it is managed by the module or not.

For the content of the attached route table, yes it is critical to be able to control it, but that's more related to the route tables custom routes feature and should be kept out of this PR. In the near future, the vcn module will also introduce full capability to create/manage route tables, not only the `internet_route` and `nat_route` route tables automatically created with the gateways.

I would suggest you create a separate PR to add LPG as a valid destination for custom_route_rules.

Comment (slmjy) [reply vcn_gateways.tf:213]: Okay, I may be will open a PR, cause I have the core code. I don't recall any use cases of LPG without subnet routing tables support and the routing table attached to LPG can only route to DRG. Maybe I'm missing some knowledge here...

Comment (kral2) [reply vcn_gateways.tf:213]: You are correct, if traffic is not routed from subnets to LPG nor it is used with a DRG, there is no use for this LPG.

Creating a route table and modifying its content may or may not be something you want to be managed by this module or even by Terraform.

Here, it is only about breaking down the end-to-end use cases (hub-and-spoke, 1:1 vcn peering) into smaller problems. First, allow creation of one or many LPG, then iterate to solve the adjacent problems (how do we route traffic to the LPG, etc ...).

Comment (hyder) [reply vcn_gateways.tf:215]: +1 for submitting this as a separate pr

**Review by hyder — CHANGES_REQUESTED** [review]
In the root terraform.tfvars.example, you need to provide a commented example of how these 2 variables will look like:

- local_peering_gateways
- internet_gateway_route_rules
- nat_gateway_route_rules

In terraformoptions doc, you also need to add for:
- local_peering_gateways

**Review by hyder — APPROVED** [review]

**Review by hyder — APPROVED** [review]
approved. some minor grammatical mistakes in the docs but we can fix that later

**Pull request merged**